### ISSUE TITLE
tf.contrib.data.rejection_resample not balancing class/freq on random crops

### ISSUE THREAD

**Issue by lrigazio** [tensorflow/tensorflow#16606]
randomly sampling / cropping data seems to break rejection_resample ==> meaning it won't do any re balancing of the class probability -- see these two simple feeders as example:

The first randomly samples data with tf.random_uniform and breaks rejection_resample -- the second with static random data having  the same distribution is correctly resampled (output has p(class=0)=0)

This creates issues when trying to real-time sample/crop/data-augment and at the same time rebalance classes with on the same pipeline

```python
def get_data_breaks(self, batch_size, iihook, this_set='train'):
    def sample(data, label): # sample detection window inside chunk
        xx = tf.cast(tf.random_uniform([1])*self.class_num, tf.int32)[0]
        with tf.control_dependencies([xx]): tf.Print(xx , [xx], 'xx>>')
        return xx, xx

    initial_dist=[1.0/self.class_num for cc in range(self.class_num)]
    classes = np.random.choice(self.class_num,20000,p=initial_dist)

    data_ph = tf.placeholder(classes.dtype, classes.shape)
    labels_ph = tf.placeholder(classes.dtype, classes.shape)
    dataset = tf.data.Dataset.from_tensor_slices((data_ph, labels_ph))

    dataset = dataset.map(sample, num_parallel_calls=1)

    target_dist=[1.0/self.class_num for cc in range(self.class_num)]
    target_dist[1]+=target_dist[0] ; target_dist[0]=0
    print('target-dist>>', target_dist)
    initial_dist = None

    dataset = dataset.apply(tf.contrib.data.rejection_resample(
                class_func=lambda c, _: c,
                target_dist=target_dist,
                initial_dist=initial_dist,
                seed=42)).map(lambda a,b: b)

    dataset = dataset.repeat(None)
    iterator = dataset.make_initializable_iterator()
    iihook.iterator_initializer_func = lambda sess: sess.run(iterator.initializer,
                    feed_dict={data_ph: classes, labels_ph: classes})

    return iterator.get_next()


def get_data_works(self, batch_size, iihook, this_set='train'):

    initial_dist=[1.0/self.class_num for cc in range(self.class_num)]
    classes = np.random.choice(self.class_num,20000,p=initial_dist)

    data_ph = tf.placeholder(classes.dtype, classes.shape)
    labels_ph = tf.placeholder(classes.dtype, classes.shape)
    dataset = tf.data.Dataset.from_tensor_slices((data_ph, labels_ph))

    target_dist=[1.0/self.class_num for cc in range(self.class_num)]
    target_dist[1]+=target_dist[0] ; target_dist[0]=0
    print('target-dist>>', target_dist)
    initial_dist = None

    dataset = dataset.apply(tf.contrib.data.rejection_resample(
                class_func=lambda c, _: c,
                target_dist=target_dist,
                initial_dist=initial_dist,
                seed=42)).map(lambda a,b: b)

    dataset = dataset.repeat(None)
    iterator = dataset.make_initializable_iterator()
    iihook.iterator_initializer_func = lambda sess: sess.run(iterator.initializer,
                    feed_dict={data_ph: classes, labels_ph: classes})

    return iterator.get_next()

**Comment by ebrevdo** [comment]
@joel-shor ptal?

**Comment by lrigazio** [comment]
I must say my use-case is a bit different ~ maybe that's why we didn't run into this issue before.

Basically my sampler can change class_it (this is just a skeleton of my main sampler // minimal code that breaks) --  think of sampling random part of an image and depending where you land you hit a different object so class_id will be different ... that clearly will change stats for the following rejection_resampler ~ but somehow the resampler bails and gives me uniform (input?) frequencies when I use the TF sampler (call the map(sample)) - but works when using fixed random stream

I can post a fully runnable script if you wish / if it helps

ps// hello png :)

**Comment by lrigazio** [comment]
here, I made it nice and self-contained -- just run this --

```python
import numpy as np
import tensorflow as tf

def fix_sample_and_rebalance(which='works'):
    class_num=8
    def get_data_breaks(batch_size=1):
        def sample(data, label): # sample detection window inside chunk
            xx = tf.cast(tf.random_uniform([1])*class_num, tf.int32)[0]
            # with tf.control_dependencies([xx]): tf.Print(xx , [xx], 'xx>>')
            return xx, xx

        initial_dist=[1.0/class_num for cc in range(class_num)]
        classes = np.random.choice(class_num,20000,p=initial_dist)

        data_ph = tf.placeholder(classes.dtype, classes.shape)
        labels_ph = tf.placeholder(classes.dtype, classes.shape)
        dataset = tf.data.Dataset.from_tensor_slices((data_ph, labels_ph))

        dataset = dataset.map(sample, num_parallel_calls=1)

        target_dist=[1.0/class_num for cc in range(class_num)]
        target_dist[1]+=target_dist[0] ; target_dist[0]=0
        print('target-dist>>', target_dist)
        initial_dist = None

        dataset = dataset.apply(tf.contrib.data.rejection_resample(
                    class_func=lambda c, _: c,
                    target_dist=target_dist,
                    initial_dist=initial_dist,
                    seed=42)).map(lambda a,b: b)

        dataset = dataset.repeat(None)
        iterator = dataset.make_initializable_iterator()
        iterator_initializer_func = lambda sess: sess.run(iterator.initializer,
                        feed_dict={data_ph: classes, labels_ph: classes})

        return iterator.get_next(), iterator_initializer_func

    def get_data_works(batch_size=1):
        initial_dist=[1.0/class_num for cc in range(class_num)]
        classes = np.random.choice(class_num,20000,p=initial_dist)

        data_ph = tf.placeholder(classes.dtype, classes.shape)
        labels_ph = tf.placeholder(classes.dtype, classes.shape)
        dataset = tf.data.Dataset.from_tensor_slices((data_ph, labels_ph))

        target_dist=[1.0/class_num for cc in range(class_num)]
        target_dist[1]+=target_dist[0] ; target_dist[0]=0
        print('target-dist>>', target_dist)
        initial_dist = None

        dataset = dataset.apply(tf.contrib.data.rejection_resample(
                    class_func=lambda c, _: c,
                    target_dist=target_dist,
                    initial_dist=initial_dist,
                    seed=42)).map(lambda a,b: b)

        dataset = dataset.repeat(None)
        iterator = dataset.make_initializable_iterator()
        iterator_initializer_func = lambda sess: sess.run(iterator.initializer,
                        feed_dict={data_ph: classes, labels_ph: classes})

        return iterator.get_next(), iterator_initializer_func


    with tf.Session() as sess:
        if which == 'works':
            get_next, iterator_initializer_func = get_data_works()
        else:
            get_next, iterator_initializer_func = get_data_breaks()
        iterator_initializer_func(sess)
        returned = []
        for kk in range(0,10000):
            try:
                sample=sess.run(get_next)
                # print(sample[0].shape, sample[1].shape, sample[1])
                returned.append(sample[0])
                print(np.bincount(np.array(returned)))
            except tf.errors.OutOfRangeError:
                break

# output is nice and (un)balanced 0 on first class 2x on second and then all unif
# [   0 2485 1284 1267 1271 1228 1250 1215]
fix_sample_and_rebalance()

# output shows po(x)==pi(x) -- rejection_resample didn't do nada~niente~nulla!
# [1193 1247 1267 1341 1264 1260 1191 1237]
fix_sample_and_rebalance('breaks')

**Comment by joel-shor** [comment]
FWIW, I'm still debugging but simplified your example:

```
import numpy as np
import tensorflow as tf

class_num=8
data_np = np.random.choice(class_num,20000,p=[1.0/class_num]*class_num)

def sample(_):
  xx = tf.cast(tf.random_uniform([1])*class_num, tf.int32)[0]
  return xx

def fix_sample_and_rebalance(which='works'): 
  data_tensors = tf.constant(data_np, dtype=tf.int32)
  dataset = tf.data.Dataset.from_tensor_slices(data_tensors)

  target_dist = [1.0/class_num] * class_num
  target_dist[1]+=target_dist[0] ; target_dist[0]=0
  print('target-dist>>', target_dist)

  if which == 'breaks':
    dataset = dataset.map(sample)

  dataset = dataset.apply(tf.contrib.data.rejection_resample(
      class_func=lambda c: c,
      target_dist=target_dist))
  dataset = dataset.map(lambda a, _: a)

  return dataset.make_one_shot_iterator().get_next()

def run_thing(which):
  tf.reset_default_graph()
  with tf.Session() as sess:
    get_next = fix_sample_and_rebalance(which)
    returned = []
    for kk in range(0,100):
      try:
          sample=sess.run(get_next)
          returned.append(sample)
      except tf.errors.OutOfRangeError:
          break
    print(np.bincount(np.array(returned)))


run_thing('works')  # this works
run_thing('breaks')  # this doesn't
```

**Comment by tensorflowbutler** [comment]
Nagging Assignee @joel-shor: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @joel-shor: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.
